### PR TITLE
Use enumset for the pseudo headers checks

### LINKED ISSUE #1273: Use enumset for the pseudo headers checks
I might prefer [enumset](https://docs.rs/enumset/1.0.8/enumset/).

_Originally posted by @martinthomson in https://github.com/mozilla/neqo/pull/1270#discussion_r755529790_

### PR DESCRIPTION
**Pull request by KershawChang**
This will fix #1273.

### CODE REVIEW

File: neqo-http3/src/headers_checks.rs [hunk 1]
```diff
@@ -5,15 +5,19 @@
...
-const REGULAR_HEADER: u8 = 0x80;
+#[derive(EnumSetType, Debug)]
+enum PseudoHeaderState {
+    Status,
+    Method,
+    Scheme,
+    Authority,
+    Path,
+    Protocol,
+    None,
```
Comment (martinthomson) [inline neqo-http3/src/headers_checks.rs:19]: Try

```suggestion
    Regular,
```

Or something like that.

File: neqo-http3/src/headers_checks.rs [hunk 2]
```diff
@@ -30,27 +34,31 @@ pub fn is_interim(headers: &[Header]) -> Res<bool> {
...
-            (MessageType::Request, ":scheme") => PSEUDO_HEADER_SCHEME,
-            (MessageType::Request, ":authority") => PSEUDO_HEADER_AUTHORITY,
-            (MessageType::Request, ":path") => PSEUDO_HEADER_PATH,
-            (MessageType::Request, ":protocol") => PSEUDO_HEADER_PROTOCOL,
+            (MessageType::Response, ":status") => PseudoHeaderState::Status,
+            (MessageType::Request, ":method") => PseudoHeaderState::Method,
+            (MessageType::Request, ":scheme") => PseudoHeaderState::Scheme,
+            (MessageType::Request, ":authority") => PseudoHeaderState::Authority,
+            (MessageType::Request, ":path") => PseudoHeaderState::Path,
+            (MessageType::Request, ":protocol") => PseudoHeaderState::Protocol,
```
Comment (martinthomson) [inline neqo-http3/src/headers_checks.rs:52]: Maybe implement `TryFrom<(MessageType, &str)>` for the enum.

File: neqo-http3/src/headers_checks.rs [hunk 3]
```diff
@@ -30,27 +34,31 @@ pub fn is_interim(headers: &[Header]) -> Res<bool> {
...
         };
         (true, bit)
     } else {
-        (false, REGULAR_HEADER)
+        (false, PseudoHeaderState::None)
     };
 
-    if *state & bit == 0 || !pseudo {
-        *state |= bit;
+    if !state.contains(bit) || !pseudo {
```
Comment (martinthomson) [inline neqo-http3/src/headers_checks.rs:60]: pseudo can be an attribute on the enum now.

**Review by martinthomson — COMMENTED** [review]
Seems like a good move.  Some refinement might be in order though.

File: neqo-http3/src/headers_checks.rs (hunk 2 above)
Comment (KershawChang) [reply neqo-http3/src/headers_checks.rs:52]: Nice suggestion. Thanks.

**Review by martinthomson — COMMENTED** [review]
r+ with changes as noted.

**Comment by KershawChang** [comment]
I am not sure what to do with the clippy error below.
```
error: you are implementing `Clone` explicitly on a `Copy` type
  --> neqo-http3/src/headers_checks.rs:13:10
   |
13 | #[derive(EnumSetType, Debug)]
   |          ^^^^^^^^^^^
   |
note: the lint level is defined here
  --> neqo-http3/src/lib.rs:7:45
   |
7  | #![cfg_attr(feature = "deny-warnings", deny(warnings))]
   |                                             ^^^^^^^^
   = note: `#[deny(clippy::expl_impl_clone_on_copy)]` implied by `#[deny(warnings)]`
note: consider deriving `Clone` or removing `Copy`
  --> neqo-http3/src/headers_checks.rs:13:10
   |
13 | #[derive(EnumSetType, Debug)]
   |          ^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expl_impl_clone_on_copy
   = note: this error originates in the derive macro `EnumSetType` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Adding `#[allow(clippy::expl_impl_clone_on_copy)]` seems also not helpful.

**Comment by martinthomson** [comment]
I find it a little annoying that EnumSet doesn't just check that there are implementations of Clone, Copy, PartialEq, and Eq, but that's their decision really.

The warning can be suppressed by adding the suppression to the module.  The problem is that the warning suppression annotation is not associated with the generated code:

```diff
diff --git a/neqo-http3/src/headers_checks.rs b/neqo-http3/src/headers_checks.rs
index 2d408718..700bcbd1 100644
--- a/neqo-http3/src/headers_checks.rs
+++ b/neqo-http3/src/headers_checks.rs
@@ -1,22 +1,23 @@
 // Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
 // http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
 // <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+#![allow(clippy::expl_impl_clone_on_copy)] // see 
+
 use crate::{Error, MessageType, Res};
 use enumset::{enum_set, EnumSet, EnumSetType};
 use neqo_common::Header;
 use std::convert::TryFrom;
 
-#[allow(clippy::expl_impl_clone_on_copy)]
 #[derive(EnumSetType, Debug)]
 enum PseudoHeaderState {
     Status,
     Method,
     Scheme,
     Authority,
     Path,
     Protocol,
     Regular,
 }
```

**Review by martinthomson — APPROVED** [review]

**Pull request merged**